Profiling allocator for mobile.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43951 Profiling allocator for mobile.**

Summary:
AllocationPlan: 
Stores the sequence of allocations, their sizes and liftime of the allocations. Along with this it also stores the total size of a single memory blob, total_size, required to satisfy all the allocations. It also stores the offsets in the blob, of size total_size, corresponding to each allocation.
Thus allocation plan contains:
- allocation sizes
- allocation lifetimes
- allocation offsets
- total size
AllocationPlaner: Takes a pointer to the allocation plan and fills it ups with plan, i.e. sizes, lifetimes, offsets, total size. This is done via WithProfileAllocationsGuard which takes in AllocationPlan* and constructs AllocationPlanner* and set the thread local allocation_planner to it. MobileCPUAllocator profiles allocations via allocation_planner. In WithValidateAllocationsGuard, allocations profiled in the allocation plan are validated.

CPUProfilingAllocator:
Application owns CPUProfilingAllocator Using WithProfilingAllocatorGuard, it passes both CPUProfilingAllocator and AllocationPlan created earlier. Then CPUProfilingAllocator will manage allocations and frees according to the plan. Allocations that are not managed by CPUProfilingAllocator will be routed through c10::alloc_cpu, c10::free_cpu.

Test Plan:
cpu_profiling_allocator_test on mobile.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23451019](https://our.internmc.facebook.com/intern/diff/D23451019)

kimishpatel committed:
Profiling allocator for mobile.

Summary:
AllocationPlan: Stores the sequence of allocations, their sizes
                and liftime of the allocations. Along with this
                it also stores the total size of a single memory
                blob, total_size, required to satisfy all the allocations.
                It also stores the offsets in the blob, of size
                total_size, corresponding to each allocation.
                Thus allocation plan contains:
                - allocation sizes
                - allocation lifetimes
                - allocation offsets
                - total size
AllocationPlaner: Takes a pointer to the allocation plan and fills
                  it ups with plan, i.e. sizes, lifetimes, offsets,
                  total size.
                  This is done via WithProfileAllocationsGuard which
                  takes in AllocationPlan* and constructs
                  AllocationPlanner* and set the thread local
                  allocation_planner to it.
                  MobileCPUAllocator profiles allocations via
                  allocation_planner.
                  In WithValidateAllocationsGuard, allocations profiled
                  in the allocation plan are validated.
CPUProfilingAllocator:
Application owns CPUProfilingAllocator
Using WithProfilingAllocatorGuard, it passes both CPUProfilingAllocator
and AllocationPlan created earlier. Then CPUProfilingAllocator will
manage allocations and frees according to the plan. Allocations that
are not managed by CPUProfilingAllocator will be routed through
c10::alloc_cpu, c10::free_cpu.

Test Plan:
cpu_profiling_allocator_test on mobile.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on c10/core/CPUProfilingAllocator.h:
```diff
@@ -0,0 +1,146 @@
+ * {
+ *   WithValidateAllocationPlanGaurd(&plan, &plan_validation_success);
+ *   module.forward(...);
+ *   success = success && plan_validation_success;
```
Mainly so that user can decide note to use ProfilingAllocator based on the result of success. Other option is to have user run that in try.. catch block and make the same decision based on the outcome.

ezyang replied:
Exceptions really slow, so if you expect this to be failing on a regular basis, bool return is much better.

kimishpatel commented on c10/core/CPUProfilingAllocator.h:
```diff
@@ -0,0 +1,146 @@
+class C10_API WithProfilingAllocatorGuard {
+  public:
+    WithProfilingAllocatorGuard(
+        CPUProfilingAllocator* allocator, AllocationPlan* const plan);
```
Oh yes. I got it reversed.

kimishpatel commented on c10/core/CPUProfilingAllocator.cpp:
```diff
@@ -0,0 +1,404 @@
+}
+
+void* CPUProfilingAllocator::allocate(const size_t bytes) {
+  TORCH_CHECK(bytes == plan_->allocation_sizes[allocation_id_],
+      "Got allocation request that does not match with the plan.");
```
I think this makes sense. We can actually do padded profiling allocator, where deploying profiling allocator allocates padded # of more bytes. However the real problem is elsewhere. In XNNPACK we reallocate input if allocator was not MobileCPUAllocator, this means we get a different sequence of allocation requests on server vs mobile. We can potentially figure it out but I think a better option would be to provide a way to use the same allocator that we will use when deploying ProfilingAllocator. So in this case can do via https://github.com/pytorch/pytorch/blob/master/c10/core/CPUAllocator.cpp#L207, something like `SetCPUAllocator(GetDefaultMobileCPUAllocator())`.

kimishpatel commented on c10/core/CPUProfilingAllocator.h:
```diff
@@ -0,0 +1,146 @@
+    uint64_t allocation_id_{0};
+    uint64_t current_size_{0};
+    void* blob_{nullptr};
+    ska::flat_hash_map<const void*, uint64_t> allocation_ptr_to_id_;
```
Right. What is not managed by ProfilingAllocator can only be identified by the lack of its presence in this map for which ptr needs to be looked up.

kimishpatel commented on c10/core/CPUProfilingAllocator.cpp:
```diff
@@ -0,0 +1,404 @@
+    //    for (....) {
+    //      {
+    //        CPUProfilingAllocator
+    //        out = ...some op (This also previous memory help by out)
```
That comment should read "This also frees previous memory held by out".
In that case lifetime of out is not managed by profiling allocator as lifetime spans beyond its scope. However both allocation and release of out, both, are observed inside the scope.

kimishpatel commented on c10/core/CPUProfilingAllocator.cpp:
```diff
@@ -0,0 +1,404 @@
+}
+
+CPUProfilingAllocator::~CPUProfilingAllocator() {
+  c10::free_cpu(blob_);
```
I think precise check might be expensive as we will have to record allocation_id's of everything that is freed. We can use a bitmask and check it against what it is supposed to be. That does not sound that expensive though.
On the other hand imprecise check, which I think should be fairly accurate, might be easier. We just count the number of `free` that we saw and did not do `c10::free_cpu` for. This should match the number of allocations we manage.
But then throwing from destructor is an issue.

kimishpatel commented on c10/core/CPUProfilingAllocator.cpp:
```diff
@@ -0,0 +1,404 @@
+  // Nesting of allocation profiling does not seem meanigful.
+  TORCH_CHECK(allocation_planner == nullptr,
+      "Nesting profiling allocations is not supported.");
+  planner_ = std::make_unique<AllocationPlanner>(plan);
```
Out of habit probably. I will fix it.

kimishpatel commented on c10/core/CPUProfilingAllocator.cpp:
```diff
@@ -0,0 +1,404 @@
+    const uint64_t size, const void* ptr) {
+  if (allocation_id_ >= allocation_plan_->allocation_sizes.size() ||
+      allocation_plan_->allocation_sizes[allocation_id_] != size) {
+    return false;
```
So for the usage I have in mind, I still want to return boolean. Let me add warning with details.

kimishpatel committed:
Update on "Profiling allocator for mobile."


Summary:
AllocationPlan: 
Stores the sequence of allocations, their sizes and liftime of the allocations. Along with this it also stores the total size of a single memory blob, total_size, required to satisfy all the allocations. It also stores the offsets in the blob, of size total_size, corresponding to each allocation.
Thus allocation plan contains:
- allocation sizes
- allocation lifetimes
- allocation offsets
- total size
AllocationPlaner: Takes a pointer to the allocation plan and fills it ups with plan, i.e. sizes, lifetimes, offsets, total size. This is done via WithProfileAllocationsGuard which takes in AllocationPlan* and constructs AllocationPlanner* and set the thread local allocation_planner to it. MobileCPUAllocator profiles allocations via allocation_planner. In WithValidateAllocationsGuard, allocations profiled in the allocation plan are validated.

CPUProfilingAllocator:
Application owns CPUProfilingAllocator Using WithProfilingAllocatorGuard, it passes both CPUProfilingAllocator and AllocationPlan created earlier. Then CPUProfilingAllocator will manage allocations and frees according to the plan. Allocations that are not managed by CPUProfilingAllocator will be routed through c10::alloc_cpu, c10::free_cpu.

Test Plan:
cpu_profiling_allocator_test on mobile.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23451019](https://our.internmc.facebook.com/intern/diff/D23451019)

[ghstack-poisoned]

kimishpatel committed:
Update on "Profiling allocator for mobile."


Summary:
AllocationPlan: 
Stores the sequence of allocations, their sizes and liftime of the allocations. Along with this it also stores the total size of a single memory blob, total_size, required to satisfy all the allocations. It also stores the offsets in the blob, of size total_size, corresponding to each allocation.
Thus allocation plan contains:
- allocation sizes
- allocation lifetimes
- allocation offsets
- total size
AllocationPlaner: Takes a pointer to the allocation plan and fills it ups with plan, i.e. sizes, lifetimes, offsets, total size. This is done via WithProfileAllocationsGuard which takes in AllocationPlan* and constructs AllocationPlanner* and set the thread local allocation_planner to it. MobileCPUAllocator profiles allocations via allocation_planner. In WithValidateAllocationsGuard, allocations profiled in the allocation plan are validated.

CPUProfilingAllocator:
Application owns CPUProfilingAllocator Using WithProfilingAllocatorGuard, it passes both CPUProfilingAllocator and AllocationPlan created earlier. Then CPUProfilingAllocator will manage allocations and frees according to the plan. Allocations that are not managed by CPUProfilingAllocator will be routed through c10::alloc_cpu, c10::free_cpu.

Test Plan:
cpu_profiling_allocator_test on mobile.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23451019](https://our.internmc.facebook.com/intern/diff/D23451019)

[ghstack-poisoned]

kimishpatel committed:
Update on "Profiling allocator for mobile."


Summary:
AllocationPlan: 
Stores the sequence of allocations, their sizes and liftime of the allocations. Along with this it also stores the total size of a single memory blob, total_size, required to satisfy all the allocations. It also stores the offsets in the blob, of size total_size, corresponding to each allocation.
Thus allocation plan contains:
- allocation sizes
- allocation lifetimes
- allocation offsets
- total size
AllocationPlaner: Takes a pointer to the allocation plan and fills it ups with plan, i.e. sizes, lifetimes, offsets, total size. This is done via WithProfileAllocationsGuard which takes in AllocationPlan* and constructs AllocationPlanner* and set the thread local allocation_planner to it. MobileCPUAllocator profiles allocations via allocation_planner. In WithValidateAllocationsGuard, allocations profiled in the allocation plan are validated.

CPUProfilingAllocator:
Application owns CPUProfilingAllocator Using WithProfilingAllocatorGuard, it passes both CPUProfilingAllocator and AllocationPlan created earlier. Then CPUProfilingAllocator will manage allocations and frees according to the plan. Allocations that are not managed by CPUProfilingAllocator will be routed through c10::alloc_cpu, c10::free_cpu.

Test Plan:
cpu_profiling_allocator_test on mobile.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23451019](https://our.internmc.facebook.com/intern/diff/D23451019)

[ghstack-poisoned]

ezyang commented on c10/core/CPUAllocator.cpp:
```diff
@@ -156,13 +157,20 @@ class DefaultMobileCPUAllocator final : public at::Allocator {
     // TODO: enable with better TLS support on mobile
     // profiledCPUMemoryReporter().Delete(pointer);
     auto allocator_ptr = GetThreadLocalCachingAllocator();
+    auto profiling_allocator_ptr = GetThreadLocalProfilingAllocator();
```
Is there any particular reason this lives as a separate concept from caching allocator? Given that there is never a situation where BOTH allocators are used, it seems like having two separate thread local allocators is unnecessary?

ezyang replied:
By the way, the use of thread local allocator here seems super questionable. If pointers escape from the thread local scope of the allocator, you will deallocate them with the wrong allocator here (which will probably cause a segfault). And we already have machinery for making sure you associate a correct deleter with each allocation (that's what DataPtr is for), so I'm not sure why you have to do it this way.

kimishpatel replied:
> Is there any particular reason this lives as a separate concept from caching allocator? Given that there is never a situation where BOTH allocators are used, it seems like having two separate thread local allocators is unnecessary?

Thanks Ed for the comment. Idea with profiling allocator is that you create a plan offline and use that plan at runtime, throwing exception when it fails. This PR introduces only the first part, where allocation plan does not have to be serialized with model but later we want to introduce those changes so as to not have to run profiling and allocation planning on mobile. Caching allocator on the other hand is model independent and it allows user to implements its own allocator by making caching allocator a virtual class.

kimishpatel replied:
> If pointers escape from the thread local scope of the allocator

Can you explain what do you mean by this?

If you mean that pointer is allocated within the thread local scope but deallocated outside, the correct behavior should be that such an allocation is not managed by profiling allocator. If there is a bug that does not ensure that behavior we need to fix that, but the design is that only allocations that are allocated and freed within this thread local scope are managed by profiling allocator.

ezyang replied:
> This PR introduces only the first part, where allocation plan does not have to be serialized with model but later we want to introduce those changes so as to not have to run profiling and allocation planning on mobile.

OK, I'll trust you on this. My general point was that if you only have one thread local allocator, you might be able to manage this at a higher level, in the selection of what allocator you register to the thread local at any given point. But maybe this is not so easy to do in the future.

ezyang commented on c10/core/CPUAllocator.cpp:
```diff
@@ -179,10 +187,17 @@ class DefaultMobileCPUAllocator final : public at::Allocator {
     if (allocator_ptr != nullptr) {
       data = allocator_ptr->allocate(alloc_size);
+    } else if (profiling_allocator_ptr != nullptr) {
+      data = profiling_allocator_ptr->allocate(alloc_size);
```
Continuing from https://github.com/pytorch/pytorch/pull/43951/files#r492446863 the way to fix this problem is to change the API of the `allocate` indirection here to return a DataPtr instead of a `void*` pointer. Then the thread local profiling allocator that was in scope at this time can provide an appropriate deleter (rather than having to do the lookup again upon deletion). You can then dispense with the `deleter` static method entirely.

kimishpatel replied:
So as I said in my previous comments profiling allocator manages only those allocations which are allocated + freed within the scope. However, your point might still be valid, so I need to think if this is a better way to do the same thing. I will make changes accordingly.

ezyang replied:
Yes, I can believe that under your intended usage pattern, the failure case that I had described would not happen. But if that's the precondition, you are at very least obligated to document the precondition (I don't think it is documented right now); and furthermore, the framework goes out of its way to make it easy for you to do things correctly even the more general situation when allocations escape the scope. (In some sense, the latter point makes it more important to handle this correctly, as most people will assume your described precondition doesn't need to hold for all allocators!)

kimishpatel replied:
> Yes, I can believe that under your intended usage pattern, the failure case that I had described would not happen.

To make sure we understand each other. What I mean is that if you use profiling allocator and utils to generate allocation plan and use that plan for allocation, it will not manage allocations whose lifetimes are outside the scope of the allocator. I will add comments to the effect, such that user would know that not all allocations made inside the scope are managed by profiling allocator.

> In some sense, the latter point makes it more important to handle this correctly, as most people will assume your described precondition doesn't need to hold for all allocators.

I can agree with this. But when you say "handle this correctly" do you mean that, "handle correctly even when the precondition (that allocations with lifetime outside the scope are not managed by the allocator) does not need to be held true"?

ezyang replied:
> What I mean is that if you use profiling allocator and utils to generate allocation plan and use that plan for allocation, it will not manage allocations whose lifetimes are outside the scope of the allocator.

Yes. Though, I thought I saw some comments that implied that the profiling process can also detect allocations that escape the scope (you assign them int max ending, right?)

> But when you say "handle this correctly" do you mean that, "handle correctly even when the precondition (that allocations with lifetime outside the scope are not managed by the allocator) does not need to be held true"?

Yes.

kimishpatel replied:
> Though, I thought I saw some comments that implied that the profiling process can also detect allocations that escape the scope (you assign them int max ending, right?)

That is correct.

ezyang commented on c10/core/CPUProfilingAllocator.h:
```diff
@@ -0,0 +1,146 @@
+    // Records size of each allocation by their sequential allocation ids.
+    std::vector<uint64_t> allocation_sizes;
+    // This maps one allocation id (X) to another allocation id (Y).
+    // Allocation X is alive until allocation Y. From allocation Y onwards
```
I read the comment after the addition of "Thus Y is the id of the first allocation after X is freed." I felt that dreiss's first statement, "X can be used to satisfy Y" was the clearest explanation of the lifetimes here.

kimishpatel replied:
I can add that if it helps clarify but the reason I did not was because we are not really creating aliases of allocation ids.

ezyang commented on c10/core/CPUProfilingAllocator.cpp:
```diff
@@ -0,0 +1,411 @@
+  MemBlock(uint64_t s, uint64_t e) : start_offset(s), end_offset(e) {}
+  bool operator<(const MemBlock& other) const {
+    return (start_offset < other.start_offset &&
+        end_offset <= other.start_offset);
```
Shouldn't a block be considered non-overlapping if it is zero size, even if `start_offset == other.start_offset`? What sense is this `<` overload being used?

ezyang replied:
If this is true, then the test here can be as simple as `end_offset <= other.start_offset`, right? (Assuming that `start_offset <= end_offset`)

kimishpatel replied:
So this is used to make sure that allocation plan has not allocated overlapping regions. I am doing this by creating binary tree in which insertions are made. Two allocations are considered overlapping and thus will return false for `<` if when comparison is made for insertion.
I think your observation is correct. Thanks for commenting. I will fix that.

kimishpatel committed:
Update on "Profiling allocator for mobile."


Summary:
AllocationPlan: 
Stores the sequence of allocations, their sizes and liftime of the allocations. Along with this it also stores the total size of a single memory blob, total_size, required to satisfy all the allocations. It also stores the offsets in the blob, of size total_size, corresponding to each allocation.
Thus allocation plan contains:
- allocation sizes
- allocation lifetimes
- allocation offsets
- total size
AllocationPlaner: Takes a pointer to the allocation plan and fills it ups with plan, i.e. sizes, lifetimes, offsets, total size. This is done via WithProfileAllocationsGuard which takes in AllocationPlan* and constructs AllocationPlanner* and set the thread local allocation_planner to it. MobileCPUAllocator profiles allocations via allocation_planner. In WithValidateAllocationsGuard, allocations profiled in the allocation plan are validated.

CPUProfilingAllocator:
Application owns CPUProfilingAllocator Using WithProfilingAllocatorGuard, it passes both CPUProfilingAllocator and AllocationPlan created earlier. Then CPUProfilingAllocator will manage allocations and frees according to the plan. Allocations that are not managed by CPUProfilingAllocator will be routed through c10::alloc_cpu, c10::free_cpu.

Test Plan:
cpu_profiling_allocator_test on mobile.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23451019](https://our.internmc.facebook.com/intern/diff/D23451019)

[ghstack-poisoned]

kimishpatel committed:
Update on "Profiling allocator for mobile."


Summary:
AllocationPlan: 
Stores the sequence of allocations, their sizes and liftime of the allocations. Along with this it also stores the total size of a single memory blob, total_size, required to satisfy all the allocations. It also stores the offsets in the blob, of size total_size, corresponding to each allocation.
Thus allocation plan contains:
- allocation sizes
- allocation lifetimes
- allocation offsets
- total size
AllocationPlaner: Takes a pointer to the allocation plan and fills it ups with plan, i.e. sizes, lifetimes, offsets, total size. This is done via WithProfileAllocationsGuard which takes in AllocationPlan* and constructs AllocationPlanner* and set the thread local allocation_planner to it. MobileCPUAllocator profiles allocations via allocation_planner. In WithValidateAllocationsGuard, allocations profiled in the allocation plan are validated.

CPUProfilingAllocator:
Application owns CPUProfilingAllocator Using WithProfilingAllocatorGuard, it passes both CPUProfilingAllocator and AllocationPlan created earlier. Then CPUProfilingAllocator will manage allocations and frees according to the plan. Allocations that are not managed by CPUProfilingAllocator will be routed through c10::alloc_cpu, c10::free_cpu.

Test Plan:
cpu_profiling_allocator_test on mobile.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23451019](https://our.internmc.facebook.com/intern/diff/D23451019)

[ghstack-poisoned]

kimishpatel committed:
Update on "Profiling allocator for mobile."


Summary:
AllocationPlan: 
Stores the sequence of allocations, their sizes and liftime of the allocations. Along with this it also stores the total size of a single memory blob, total_size, required to satisfy all the allocations. It also stores the offsets in the blob, of size total_size, corresponding to each allocation.
Thus allocation plan contains:
- allocation sizes
- allocation lifetimes
- allocation offsets
- total size
AllocationPlaner: Takes a pointer to the allocation plan and fills it ups with plan, i.e. sizes, lifetimes, offsets, total size. This is done via WithProfileAllocationsGuard which takes in AllocationPlan* and constructs AllocationPlanner* and set the thread local allocation_planner to it. MobileCPUAllocator profiles allocations via allocation_planner. In WithValidateAllocationsGuard, allocations profiled in the allocation plan are validated.

CPUProfilingAllocator:
Application owns CPUProfilingAllocator Using WithProfilingAllocatorGuard, it passes both CPUProfilingAllocator and AllocationPlan created earlier. Then CPUProfilingAllocator will manage allocations and frees according to the plan. Allocations that are not managed by CPUProfilingAllocator will be routed through c10::alloc_cpu, c10::free_cpu.

Test Plan:
cpu_profiling_allocator_test on mobile.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23451019](https://our.internmc.facebook.com/intern/diff/D23451019)

[ghstack-poisoned]

dzhulgakov commented on c10/mobile/CPUProfilingAllocator.cpp:
```diff
@@ -0,0 +1,410 @@
+    }
+    auto start_offset = allocation_offsets[i];
+    auto end_offset = allocation_offsets[i] + allocation_sizes[i];
+    if (!allocations.emplace(start_offset, end_offset).second) {
```
what exactly does this validation check? That not two allocations land in exactly the same spot? But they could if one of them got deallocated and reallocated, right? I'd imagine that you want to check actual alloc/dealloc across time instead

kimishpatel replied:
Thanks for the catch @dzhulgakov. I thought I covered this but clearly not. I will fix it.

dzhulgakov commented on c10/mobile/CPUProfilingAllocator.h:
```diff
@@ -0,0 +1,149 @@
+class C10_API AllocationPlan {
+  private:
+    // Records size of each allocation by their sequential allocation ids.
+    std::vector<uint64_t> allocation_sizes;
```
nit for future: in general vector of structs is usually more readable than a bunch of vectors
